src/Stage.java

```diff
@@ -6,18 +6,20 @@ public class Stage {
     private int pc;            // Program counter
     private int nextPc;        // Next program counter (automatically set to Pc+1)
     private boolean isHalt;    // Halt flag
+    private Decoder decoder;
 
     public Stage() {
         memory = new int[65536];
         register = new int[8];
         pc = 0;
         nextPc = 0;
         isHalt = false;
+        decoder = Decoder.getInstance();
     }
```

Comment:
Please add another field `stepCount` that dictates how many steps have been iterated. `stepCount` should be incremented at the end of `iterate()` method.